### PR TITLE
chore: [FX-2739] Add madge circularity check

### PR DESCRIPTION
**Pull request by denieler**
[FX-2739]

### Description

Add circularity check via [Madge](https://github.com/pahen/madge)

### How to test

- `yarn install`
- `yarn circularity`

<details>
<summary>PR commands</summary>
<br />

List of available commands:

- `@toptal-bot run all` - Run whole pipeline
- `@toptal-bot run build` - Check build
- `@toptal-bot run visual` - Run visual tests
- `@toptal-bot run deploy:documentation` - Deploy documentation
- `@toptal-bot run package:alpha-release` - Release alpha version

</details>

<details>
<summary>PR Review Guidelines</summary>
<br />

#### When to approve? ✅

**You are OK** with merging this PR and

1. You have no extra requests.
2. You have optional requests.
   1. Add `nit:` to your comment. (ex. `nit: I'd rename this variable from makeCircle to getCircle`)

#### When to request changes? ❌

**You are not OK** with merging this PR because

1. Something is broken after the changes.
2. Acceptance criteria is not reached.
3. Code is dirty.

#### When to comment (neither ✅ nor ❌)

**You want your comments to be addressed** before merging this PR in cases like:

1. There are leftovers like unnecessary logs, comments, etc.
2. You have an opinionated comment regarding the code that requires a discussion.
3. You have questions.

#### How to handle the comments?

1. An owner of a comment is the only one who can resolve it.
2. An owner of a comment must resolve it when it's addressed.
3. A PR owner must reply with ✅ when a comment is addressed.

</details>

[FX-2739]: https://toptal-core.atlassian.net/browse/FX-2739?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by separatio** [comment]
Is this PR going to also fix circularity issues? I found 91 locally.

**Comment by denieler** [comment]
> Is this PR going to also fix circularity issues? I found 91 locally.

@separatio no, only adding the command to check them 😃

**Review by separatio — APPROVED** [review]
Straightforward. Looks cool. Can't wait to use it more.

File: .madgerc [hunk 1]
```diff
@@ -0,0 +1,4 @@
+{
+  "fileExtensions": ["js", "ts", "jsx", "tsx"],
+  "tsConfig": "./tsconfig.json"
```
Comment (separatio) [inline .madgerc:3]: Should we exclude `dist-package` directories via `excludeRegExp` or something else?

Comment (denieler) [reply .madgerc:3]: yeah, let me check 👍

Comment (denieler) [reply .madgerc]: @separatio done 👍 `dist-package` folders are excluded

File: .madgerc [hunk 2]
```diff
@@ -0,0 +1,5 @@
+{
+  "fileExtensions": ["js", "ts", "jsx", "tsx"],
+  "tsConfig": "./tsconfig.json",
+  "excludeRegExp": ["dist-package"]
```
Comment (TomasSlama) [inline .madgerc:4]: should we also exclude stories and tests? 🤔

Comment (separatio) [reply .madgerc:4]: Probably. I just encountered an edge case that included them.

Comment (denieler) [reply .madgerc:4]: @TomasSlama let's do that action whenever we will see the problem exists. So far I haven't seen any tests of stories in the report

**Review by TomasSlama — APPROVED** [review]

File: .madgerc (hunk 2 above)
Comment (separatio) [reply .madgerc:4]: If you run `yarn circularity -s`, which gives a summary of all files checked, you can see the stories, with 0 circularities offenses. The same goes for styles.ts.

Comment (separatio) [reply .madgerc:4]: As for tests, we actually have offenses for tests. For example, I found 4 circularity dependencies offenses in `picasso/src/NativeSelect/test.tsx`

Comment (denieler) [reply .madgerc]: ok @separatio I've checked and it looks like `excludeRegExp` option doesn't exclude files from the check, but exclude them only from the output. So it's pretty hard to properly filter files from the check in `madge`. I stopped at the current solution
```
madge --circular packages/*/src
```

it works under 6 sec, so looks like good enough 👍 if we would have circularity in our tests or stories - it will report that, but so far we didn't have any cases like that, so I would wait until that moment. Efforts to fix or remove some further files would be noticed and the code qualify and maintenance will be not good enough. So I don't think it worth it.
